File: cpp/dijkstra/dijkstra.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>
#include <stdexcept>
#include "../TDA/nodoDijkstra/NodoDijkstra.h"
#include "dijkstra.h"

std::pair<std::vector<std::tuple<int, int>>, float> dijkstra(std::vector<NodoDijkstra *> &nodos, std::tuple<int, int> vertice_origen, std::tuple<int, int> vertice_destino)
{
    NodoDijkstra *origen = nullptr;
    NodoDijkstra *destino = nullptr;
    std::cout << "FAdadaDA";
    for (auto &nodo : nodos)
    {

       std::cout << std::get<0>(nodo->getCoordenadas()) << std::get<1>(nodo->getCoordenadas()) << '\n';
        if (nodo->getCoordenadas() == vertice_origen)
        {
            origen = nodo;
        }
        if (nodo->getCoordenadas() == vertice_destino)
        {
            destino = nodo;
        }
    }
    std::cout << "Esta vaina va a acabar";
    if (!origen || !destino)
    {
        std::cout << "efectivamente";
        throw std::invalid_argument("Origen o destino no encontrados en la lista de nodos.");
    }

    origen->setDistancia(0);

    std::vector<NodoDijkstra *> unexplored = nodos;

    while (!unexplored.empty())
    {
        auto min_it = std::min_element(unexplored.begin(), unexplored.end(),
                                       [](NodoDijkstra *a, NodoDijkstra *b)
                                       { return a->getDistancia() < b->getDistancia(); });
        NodoDijkstra *v = *min_it;
        // std::cout << "Vertice usando : " << std::get<0>(v->getCoordenadas()) << " " << std::get<1>(v->getCoordenadas()) << '\n';

        // std::cout << "Adyacentes: " << '\n';
        // auto adyacentes = v->getAdyacentes();
        // for (const auto &adyacente : adyacentes)
        // {
        //     const auto &[key, value] = adyacente; // Desempaqueta el std::pair
        //     const auto &[y, x] = key;             // Desempaqueta el std::tuple<int, int>
        //     std::cout << "(" << x << ", " << y << ") -> " << value << "  " << "\n";
        // }
        unexplored.erase(min_it);

        if (v == destino)
        {
            break;
        }

        for (const auto &[coordenadas, peso] : v->getAdyacentes())
        {

            NodoDijkstra *adyacente = nullptr;
            for (auto &nodo : nodos)
            {
                if (nodo->getCoordenadas() == coordenadas)
                {
                    adyacente = nodo;
                    break;
                }
            }

            if (!adyacente)
                continue;

            float nueva_distancia = v->getDistancia() + peso;
            if (nueva_distancia < adyacente->getDistancia())
            {
                adyacente->setDistancia(nueva_distancia);
                adyacente->setPredecesor(v);
            }
        }
    }
    

    std::vector<std::tuple<int, int>> camino;
    NodoDijkstra *nodo_actual = destino;
    while (nodo_actual != nullptr)
    {

        auto coordenadas = nodo_actual->getCoordenadas();

        camino.insert(camino.begin(), coordenadas);
        nodo_actual = nodo_actual->getPredecesor();
        // if (nodo_actual != nullptr)
        // {
        //     std::cout << "HYPERMEGA GA" << std::get<0>(nodo_actual->getCoordenadas()) << std::get<1>(nodo_actual->getCoordenadas()) << '\n';
        // }
    }

    return {camino, destino->getDistancia()};
}

File: python/dijkstra/dijkstra.py
def dijkstra(nodos, vertice_origen, vertice_destino):
    # Busca el nodo de origen y el nodo de destino por posición
    origen = next((nodo for nodo in nodos if nodo.vertice == vertice_origen), None)
    destino = next((nodo for nodo in nodos if nodo.vertice == vertice_destino), None)

    # Verifica que el origen y el destino existen en la grilla de nodos
    if origen is None or destino is None:
        raise ValueError("Origen o destino no encontrados en la lista de nodos.")
    
    # Inicializamos el origen con distancia 0
    origen.distancia = 0
    unexplored = list(nodos)  # Copia de la lista de nodos sin explorar

    while unexplored:
        # Encontrar el nodo no explorado con la menor distancia
        v = min(unexplored, key=lambda nodo: nodo.distancia)
        unexplored.remove(v)  # Marcamos a v como explorado

        # Si llegamos al destino, terminamos
        if v == destino:
            break

        # Revisar cada adyacente del vertice
        for coordenada, peso in v.adyacentes.items():
            # Buscar el nodo correspondiente a la coordenada
            adyacente = next((nodo for nodo in nodos if nodo.vertice == coordenada), None)
            if adyacente is None:
                continue  # Si no se encuentra el nodo, continuar con el siguiente adyacente
            # Halla la nueva distancia
            nueva_distancia = v.distancia + peso
            # Compara si la nueva distancia es menor a la distancia actual
            if nueva_distancia < adyacente.distancia:
                # Si la distancia es menor, esta sera reemplazada
                adyacente.distancia = nueva_distancia
                adyacente.predecesor = v

    # Reconstruir el camino desde el destino al origen
    camino = []
    nodo_actual = destino
    while nodo_actual is not None:
        # Inserta al inicio de la lista el nombre del nodo actual
        camino.insert(0, nodo_actual.vertice)
        # Obtiene el predecesor para poder seguir construyendo el camino
        nodo_actual = nodo_actual.predecesor

    return camino, destino.distancia

